Parallel demo to test ProgressHandler queue strategy

garrettmflynn opened the pull request:
This PR is a sandbox environment to test concerns raised in https://github.com/catalystneuro/tqdm_publisher/pull/49. 

While implementing this, it became apparent that we can't use the `TQDMProgressHandler` directly when working with parallel progress bars because these exist on independent threads. To overcome this while still supporting the new class, all updates coming from the HTTP relay server are registered manually to the handler with the `_announce` method.

garrettmflynn committed:
Parallel demo to test ProgressHandler queue strategy

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

CodyCBakerPhD commented:
Thanks @garrettmflynn this strategy makes a lot more sense now and the code reads better (I'll still suggest various improvements for annotations, but I'm pretty sure I follow what's going on) - and is more realistic to our use in the GUIDE to boot

Unfortunately, this is what it looks like on my end (Windows)

https://github.com/catalystneuro/tqdm_publisher/assets/51133164/9a9f6228-366f-487a-b9fa-bb0722ecc77c

I had seen this behavior a while back with the multiple bar setup, where if there were too many updates going at once the web page would freeze, so I tried slowing the rate of the tasks (which is what you see in the video; all values just multiplied by `10`). I even tried reducing down to two tasks to perform and still no dice

Especially unfortunately I cannot exit this task with `ctrl+C` `KeyboardInterrupt`, which you see in the last ~15 seconds of the video; at some point, something stops and the web page does a partial update but the only way I can kill the entire thing is to kill that terminal

I'll give it a try and see if the Mac one works

CodyCBakerPhD commented:
Yeah, the Mac one works fine for some reason

Now that I see it running, one thing I'd suggest is the very first progress bar at the top - we want to be tracking the total progress over the length of the outer list of task times (so like, 1/7 after a single job completes a full list of tasks, then 2/7, etc. each will we also have the bars below showing the progress of each individual of the 7 tasks)

https://github.com/catalystneuro/tqdm_publisher/assets/51133164/f49b0d03-2311-4fa5-933f-25675621426b

CodyCBakerPhD commented:
I wonder, could you let a Flask endpoint run that progress handler (even more like on GUIDE) and see if that perhaps helps? (It's OK for now if the flask server has to be run on a separate terminal window if that helps - or have the demo launch command be another endpoint of the server able to be controlled by a button on the client)

garrettmflynn committed:
Separate ws and non-ws implementation

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

garrettmflynn commented:
Alright I've switched the main parallel demo to Flask and added a `demo_parallel_ws` command to reference the previous implementation.

Just FYI there were two implementations in this demo: one that was _not_ queued, which was enabled by default, and one that _is_ queued. 

I've swapped this now so that the queued version (rather than direct forwarding) is enabled by default.

Can you confirm if the Windows issue is still seen on one or both demos?

CodyCBakerPhD commented:
The flask one works great!

The websocket one still has the same problems of not updating and being unable to cancel

Can you add that 'total' top bar for the number of tasks completed? (always fixed in that top position)

garrettmflynn committed:
Show global job bar

garrettmflynn committed:
Merge branch 'parallel-queue-test' of https://github.com/catalystneuro/tqdm_publisher into parallel-queue-test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

garrettmflynn commented:
Added!

garrettmflynn committed:
Update style and handle more on the client

garrettmflynn committed:
Merge branch 'parallel-queue-test' of https://github.com/catalystneuro/tqdm_publisher into parallel-queue-test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

CodyCBakerPhD commented on src/tqdm_publisher/_demos/_parallel_bars/_server.py:
```diff
@@ -0,0 +1,221 @@
+        # Send initialization for pool progress bar
+        forward_to_http_server(url, request_id, id=request_id, n=0, total=len(TASK_TIMES))
+
+        for _ in job_map:
```
This is the line you'd probably have to wrap with the global TQDM - fairly sure that as it performed the iterations in the job_map (deployment to subprocesses) it will increment as each job completes

garrettmflynn replied:
When I used `job_map`, it only updated twice instead of all five times—resulting in an incomplete bar.

CodyCBakerPhD replied:
Might need to switch to the `as_completed` strategy then using a `futures: list` that appends each `executor.submit(...)`

Any questions on that or would you like me to do it

garrettmflynn replied:
Would be great if you can handle it. I'll convert when done

CodyCBakerPhD replied:
My attempts to wrap this line with `TQDMPublisher` (using `job_map` or not) fail to deliver any updates at all

However, I did just push a version that works for me doing the manual calls to the forwarding commands - can you tinker with this more to try to see where the issue is? It just is acting as if the `update` call of the TQDM applied to the executor never gets triggered, but it definitely is in the console

garrettmflynn replied:
Fixed! Just some minor mismatches.

Now all the global tqdm bar updates are happening from the server like any other bar.

CodyCBakerPhD commented:
I see the top bar present but it never updates

garrettmflynn commented:
Hmm it's working fine on Mac for both demos here.

Since it relies on `n` and `total` being the same, there's a chance the progress bar completes without this being explicitly sent. I'd noticed this several times before,

If this persists after switching to a global TQDM, I'll add an explicit call to pass these format options.

CodyCBakerPhD committed:
working windows version for global bar

CodyCBakerPhD commented:
I also keep seeing the following like in console while Flask is running

```
127.0.0.1 - - [04/Apr/2024 23:09:50] "GET / HTTP/1.1" 404 -
```

pings about every second or so - any idea what this is? It feels like it is the thing preventing keyboard interruption

garrettmflynn committed:
Swap from 8000 to a non-dedicated port

garrettmflynn committed:
Merge branch 'parallel-queue-test' of https://github.com/catalystneuro/tqdm_publisher into parallel-queue-test

garrettmflynn commented:
> pings about every second or so - any idea what this is? It feels like it is the thing preventing keyboard interruption

This is because port 8000 is a special port, so it seems that there's some background process constantly pinging it. 

I've bumped to a different port without such an association and this goes away.

garrettmflynn committed:
Limit websocket reconnection to 3 times

garrettmflynn commented:
> pings about every second or so - any idea what this is? It feels like it is the thing preventing keyboard interruption

These are requests to reconnect the Websocket connection. Since we're using the same client for both WS and non-WS servers, the latter will result in 404 errors for these requests.

I've limited the reconnection request to 3 tries so this doesn't go on forever.

garrettmflynn committed:
Receive all job bar events from the server

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

CodyCBakerPhD commented:
Everything is working great on my end - happy to merge this as is and make improvements in follow ups

garrettmflynn commented:
Alright sounds good to me!
